File: CPP solutions/min_stack.cpp
/*No fucking idea why this will cause MLE. I think I can save a little bit more memory by making the min stack to be a tuple. Let's think about it next time.
*/
class MinStack {
public:
    void push(int x) {
        stack.push_back(x);
        if (min.empty() || x<=min.back())
        {
            min.push_back(x);
        }
    }

    void pop() {
        int x = stack.back();
        stack.pop_back();
        if (x == min.back())
        {
            min.pop_back();
        }
    }

    int top() {
        return stack.back();
    }

    int getMin() {
        return min.back();
    }
    
private:
    vector<int> stack;
    vector<int> min;
};

File: Python solutions/min_stack.py
class MinStack:
    # @param x, an integer
    def __init__(self):
        # the stack it self
        self.A = []
        self.minS=[]
    # @return an integer
    def push(self, x):
        n=len(self.A)
        if n==0:
            self.minS.append(x)
        else:
            lastmin=self.minS[-1]
            if x<=lastmin:
                self.minS.append(x)
        self.A.append(x)
    # @return nothing
    def pop(self):
        if len(self.A)>0 and self.A.pop()==self.minS[-1]:
            self.minS.pop()
    # @return an integer
    def top(self):
        return self.A[-1]
         
 
    # @return an integer
    def getMin(self):
        return self.minS[-1]